### ISSUE TITLE
Calculation of Partial Energy

### ISSUE THREAD

**Issue by yiqichenshallwetalk** [openmm/openmm#3971]
Dear developers, 

I am working on a project which conducts Monte-Carlo + molecular dynamics simulation. In the MC part, the energy difference before and after a proposed move is required. However, computing this energy difference from the whole energies is very inefficient since each MC move displaces/inserts/deletes one molecule. So I hope to get the partial energy of the system that computes interactions between the one molecule of interest and all other molecules. I have successfully done this by adding a new function **calcPartialEnergy** in **NonbondedForceImpl** and implementing it on several platforms. I wonder if there is any way to do that without changing the source code? Thank you!

**Comment by peastman** [comment]
Can you clarify exactly what you want to calculate?  How do you define the partial energy?

Based on your description, I'm guessing you're interested only in nonbonded interactions, that you're using a non-polarizable point charge force field, and that you aren't using implicit solvent.  Are those all correct?  What nonbonded method do you use?

How do you implement the insertion or deletion of molecules?

How large is your system (approximate number of atoms)?

What platform are you using?

**Comment by yiqichenshallwetalk** [comment]
> Can you clarify exactly what you want to calculate? How do you define the partial energy?

Apologies for the confusion. By partial energy I mean the nonbonded interactions between atoms in one molecule and all other atoms in the system.  For example, if a molecule has atoms 1, 2, 3, then the partial energy is the sum of interactions between ([1, 2, 3], [4, 5, 6, ...]).

> Based on your description, I'm guessing you're interested only in nonbonded interactions, that you're using a non-polarizable point charge force field, and that you aren't using implicit solvent. Are those all correct? What nonbonded method do you use?

Yes, these are all correct. I'm using Ewald/PME nonbonded mothed.

> How do you implement the insertion or deletion of molecules?

I'm using a ghost-atom method. When a molecule is deleted, its atoms' force field parameters are set to zero(ghost atoms). And insertions will scale back ghost atoms' parameters to the original values.

> How large is your system (approximate number of atoms)?

Right now I'm testing a system with 1,500 atoms. I plan to study larger systems in the future.

> What platform are you using?

I'm running these simulations on CPU right now. I also plan to run them on CUDA.

**Comment by peastman** [comment]
Doing that with PME is complicated.  If you were using a cutoff method, it would be easy.  CustomNonbondedForce supports interaction groups for computing the interactions between limited sets of particles.  But PME makes it more complicated.  It begins by spreading the charge from the particles onto a grid.  Then it computes the interaction between each pair of grid points.  There are no particles at that point.  It doesn't know that a certain fraction of the charge on a certain grid point came from a certain particle.  It's just computing the energy of a charge distribution.  There's no way to separate out the interactions between individual particles.

There are ways to calculate it, of course.  For example, you could create three Systems: one containing just the molecule, one containing just the environment, and one containing both.  By calculating the energies of all three, you could get the interaction energy.  But that's expensive, and you said your goal was to make it efficient.  I don't know any way to do that with PME.

**Comment by craabreu** [comment]
FYI, the [openmm-nonbonded-slicing](https://openmm-nonbonded-slicing.readthedocs.io/en/stable/) plugin implements a child class of `NonbondedForce` called [SlicedNonbondeForce](https://openmm-nonbonded-slicing.readthedocs.io/en/stable/pythonapi/SlicedNonbondedForce.html). Among other things, it facilitates the calculation of nonbonded interactions between subsets of particles, especially when involving PME. In @yiqichenshallwetalk's example, one can initialize a `SlicedNonbondeForce` from the original `NonbondForce` with two subsets and then call [setParticleSubset] to add particles 1, 2, and 3 to subset 1 (the remaining particles will be in subset 0 by default). Then, the following steps can be used to obtain the partial energy:

* Call [addGlobalParameter] with arguments `("lambda", 1)`: this will add a global parameter to the context with a default value of 1.
* Call [addScalingParameter] with arguments `("lambda", 0, 1, True, True)`: this will tell OpenMM that `lambda` must be used to scale both the LJ and electrostatic interactions between subsets 0 and 1.
* Call [addScalingParameterDerivative] with argument `("lambda")`: this will tell OpenMM to compute the derivative of the potential energy with respect to the global parameter `"lambda"`.
* After creating a `Context` named `context` and containing the `SlicedNonbondedForce`, execute:
```
state = context.getState(getParameterDerivatives=True)
derivatives = state.getEnergyParameterDerivatives()
partial_energy = derivatives["lambda"]
```

With two subsets and `nonbondedMethod=PME`, `SlicedNonbondedForce` needs to execute 2 FFT calculations per step. However, it does so very efficiently via batched FFTs. Usually, this is just slightly slower than a single FFT.

[setParticleSubset]:  https://openmm-nonbonded-slicing.readthedocs.io/en/stable/pythonapi/SlicedNonbondedForce.html#nonbondedslicing.SlicedNonbondedForce.setParticleSubset
[addGlobalParameter]: https://openmm-nonbonded-slicing.readthedocs.io/en/stable/pythonapi/SlicedNonbondedForce.html#nonbondedslicing.SlicedNonbondedForce.addGlobalParameter
[addScalingParameter]: https://openmm-nonbonded-slicing.readthedocs.io/en/stable/pythonapi/SlicedNonbondedForce.html#nonbondedslicing.SlicedNonbondedForce.addScalingParameter
[addScalingParameterDerivative]: https://openmm-nonbonded-slicing.readthedocs.io/en/stable/pythonapi/SlicedNonbondedForce.html#nonbondedslicing.SlicedNonbondedForce.addScalingParameterDerivative

**Comment by yiqichenshallwetalk** [comment]
Thank you Charlles! I'll check out the nonbonded-slicing plugin to see if it works. Will get back to you as soon as possible. Thanks again.

**Comment by yiqichenshallwetalk** [comment]
> state = context.getState(getParameterDerivatives=True)
> derivatives = state.getEnergyParameterDerivatives()
> partial_energy = derivatives["lambda"]

Just finished a test run on the nonbonded-slicing plugin. It can indeed compute partial energies for different slices. However, the issue is, it is not faster than direct computation of the whole energy of the system. My understanding is, this program still computes all pair-interactions in the system, but stores them separately in ```sliceEnergies```. Is it possible to tell the program to only compute interactions in a specific slice?

**Comment by peastman** [comment]
That's really cool!

> However, the issue is, it is not faster than direct computation of the whole energy of the system.

I think that's inherent in how PME works.  Computing the interaction between two subsets of atoms involves more work than computing the total energy, not less.  Have you considered using reaction field instead of PME?

**Comment by craabreu** [comment]
@peastman is correct. Computing slices of the reciprocal-space energy will always require more work than computing it in total. In the best case, computing only the internal energy of a single subset would require roughly the same amount of work. 

`SlicedNonbondedForce` is fast in what it is supposed to do: replace `NonbondedForce` in a simulation while providing extra information and flexibility. It can be used, for instance, to simulate alchemical transformations or to perform replica exchange with solute scaling (REST).

To compute only a specific slice for analysis (for simulation, you need to compute the whole thing anyway), I would follow @peastman's suggestion and use `CustomNonbondedForce` with a reaction-field model for the electrostatic part and call the `addInteractionGroup` method.

**Comment by yiqichenshallwetalk** [comment]
@peastman @craabreu Thank you for your suggestions and explanations. I will check out the reaction-field electrostatic potential model. I still have some thoughts on PME. During the MC simulations, I'm only interested in the change of energy after moving a molecule, not the actual partial energies. Therefore, I could compute the ```total reciprocal-space energy``` + ```partial real-space energy```. In theory, ```change_in_total_energy``` = ```change_in_total_reciprocal_energy``` + ```change_in_partial_realspace_energy```. This could save a significant amount of work in real-space energy calculation, right?

**Comment by peastman** [comment]
Yes, that should be possible.  If you use a fairly long cutoff, that will shift more of the work to direct space and make the benefit larger.

You can do that by using a CustomNonbondedForce with interaction groups for the direct space part, and a NonbondedForce for reciprocal space, calling `setIncludeDirectSpace(False)` on it.

**Comment by dspoel** [comment]
Not an OpenMM solution but it may be possible in OpenMM too. We did this with GROMACS by setting charges to zero selectively and recomputing the energy incl. PME. https://pubs.acs.org/doi/abs/10.1021/ct400404q
